exonum/src/events/noise/mod.rs

```diff
@@ -40,80 +41,109 @@ pub struct HandshakeParams {
+        mut self,
+        stream: S,
+    ) -> impl Future<Item = (S, Self), Error = io::Error> {
+        done(self.noise.write_handshake_msg())
+            .map_err(|e| e.into())
             .and_then(|(len, buf)| write(stream, &buf, len))
-            .and_then(|(stream, _msg)| read(stream))
-            .and_then(move |(stream, msg)| {
-                let _buf = noise.read_handshake_msg(&msg);
-                let noise = noise.into_transport_mode()?;
-                let framed = stream.framed(MessagesCodec::new(max_message_len, noise));
-                Ok(framed)
-            })
-    });
-
-    Box::new(framed)
+            .map(move |(stream, _)| (stream, self))
+    }
+
+    fn finalize<S: AsyncRead + AsyncWrite + 'static>(
+        self,
+        stream: S,
+    ) -> Result<Framed<S, MessagesCodec>, io::Error> {
+        let noise = self.noise.into_transport_mode()?;
+        let framed = stream.framed(MessagesCodec::new(self.max_message_len, noise));
+        Ok(framed)
+    }
 }
 
-fn send_handshake(stream: TcpStream, params: &HandshakeParams) -> HandshakeResult {
-    let max_message_len = params.max_message_len;
-    let mut noise = NoiseWrapper::initiator(params);
-    let framed = write_handshake_msg(&mut noise)
-        .and_then(|(len, buf)| write(stream, &buf, len))
-        .and_then(|(stream, _msg)| read(stream))
-        .and_then(move |(stream, msg)| {
-            let _buf = noise.read_handshake_msg(&msg);
-            write_handshake_msg(&mut noise)
-                .and_then(|(len, buf)| write(stream, &buf, len))
-                .and_then(move |(stream, _msg)| {
-                    let noise = noise.into_transport_mode()?;
-                    let framed = stream.framed(MessagesCodec::new(max_message_len, noise));
-                    Ok(framed)
-                })
-        });
-
-    Box::new(framed)
+impl Handshake for NoiseHandshake {
+    fn listen<S>(self, stream: S) -> HandshakeResult<S>
+    where
+        S: AsyncRead + AsyncWrite + 'static,
+    {
+        let framed = self.read_handshake_msg(stream)
+            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.finalize(stream));
+        Box::new(framed)
+    }
+
+    fn send<S>(self, stream: S) -> HandshakeResult<S>
+    where
+        S: AsyncRead + AsyncWrite + 'static,
+    {
+        let framed = self.write_handshake_msg(stream)
+            .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.finalize(stream));
+        Box::new(framed)
+    }
 }
 
-fn read(sock: TcpStream) -> Box<Future<Item = (TcpStream, Vec<u8>), Error = io::Error>> {
+fn read<S: AsyncRead + 'static>(sock: S) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
```

Comment:
It could make sense to make the names more comprehensive for `read` and `write`, as they relate to Noise handshake messages specifically. Just an idea: maybe, create a newtype `HandshakeMessage(Vec<u8>)` and implement `read` and `write` as its methods?